test/LineIntersectionTest.cpp

```diff
@@ -0,0 +1,237 @@
+    Vector2 start1(-10.0, 10000.0), end1(-10.0, 9999.9), start2(20.0, 5.0), end2(-10.0 + EPSILON, 5.0), intersection(-10.0, 5.0);
+    Line line2(start2, end2); HalfLine halfLine(start1, end1); LineSegment lineSegment1(start1, end1), lineSegment2(start2, end2);
+    std::optional<Vector2> result = halfLine.intersect(line2);
+    EXPECT_EQ(result.value(), intersection);
+    result = Line::intersect(start1, end1, start2, end2);
+    EXPECT_EQ(result.value(), intersection);
+    result = lineSegment1.intersects(lineSegment2);
+    EXPECT_FALSE(result.has_value());
+    std::vector<Vector2> result2 = lineSegment1.multiIntersect(lineSegment2);
+    EXPECT_EQ(result2.size(), 0);
+}
+
+TEST(LineIntersectionTests, NonParallel7) {
+    // Non-parallel test case a vertical line and arbitrary line, where intersection happens for the first line at t = 0 and for the second at 0 < t < 1.
+    Vector2 start1(6.0, -2.0), end1(6.0, -500.0), start2(9.0, 1.0), end2(3.0, -5.0), intersection(6.0, -2.0);
+    Line line2(start2, end2); HalfLine halfLine(start1, end1); LineSegment lineSegment1(start1, end1), lineSegment2(start2, end2);
+    std::optional<Vector2> result = halfLine.intersect(line2);
+    EXPECT_EQ(result.value(), intersection);
+    result = Line::intersect(start1, end1, start2, end2);
+    EXPECT_EQ(result.value(), intersection);
+    result = lineSegment1.intersects(lineSegment2);
+    EXPECT_EQ(result.value(), intersection);
+    std::vector<Vector2> result2 = lineSegment1.multiIntersect(lineSegment2);
+    EXPECT_EQ(result2.size(), 1);
+    EXPECT_EQ(result2[0], intersection);
+}
+
+TEST(LineIntersectionTests, Parallel1) {
+    // Parallel test case with two 'Equal' arbitrary lines.
+    Vector2 start1(-1000.0, -800.0), end1(-200.0, -400.0), start2(-200.0, -400.0), end2(-1000.0, -800.0);
+    Line line2(start2, end2); HalfLine halfLine(start1, end1); LineSegment lineSegment1(start1, end1), lineSegment2(start2, end2);
+    std::optional<Vector2> result = halfLine.intersect(line2);
+    EXPECT_FALSE(result.has_value());
+    result = Line::intersect(start1, end1, start2, end2);
+    EXPECT_FALSE(result.has_value());
+    result = lineSegment1.intersects(lineSegment2);
+    EXPECT_EQ(result.value(), start1);
+    std::vector<Vector2> result2 = lineSegment1.multiIntersect(lineSegment2);
+    EXPECT_EQ(result2.size(), 2);
+    EXPECT_EQ(result2[0], end1);
+    EXPECT_EQ(result2[1], start1);
+}
+
+TEST(LineIntersectionTests, Parallel2) {
+    // Parallel test case where the first vertical line is a subset of the second vertical line.
+    Vector2 start1(5.0, 3.0), end1(5.0, 6.0), start2(5.0, 12.0), end2(5.0, 3.0);
+    Line line2(start2, end2); HalfLine halfLine(start1, end1); LineSegment lineSegment1(start1, end1), lineSegment2(start2, end2);
+    std::optional<Vector2> result = halfLine.intersect(line2);
+    EXPECT_FALSE(result.has_value());
+    result = Line::intersect(start1, end1, start2, end2);
+    EXPECT_FALSE(result.has_value());
```

Comment:
Did you read the documentation?

Reply:
Then I disagree with you defined intersection/the function. Any user will intuitively expect this is overlapping

Reply:
With the non-static variant I would agree to change the definition. But with this one I don't agree, since based on performance it is actually quite fast.